core/src/test/scala/unit/kafka/tools/LineMessageReaderTest.scala

```diff
@@ -0,0 +1,210 @@
+
+    props.put("headers.separator", ",")
+    assertThrowsOnInvalidPatternConfig(props, "headers.delimiter and headers.key.separator may not be equal")
+
+    props.put("headers.key.separator", ",")
+    assertThrowsOnInvalidPatternConfig(props, "headers.separator and headers.key.separator may not be equal")
+  }
+
+  private def assertThrowsOnInvalidPatternConfig(props: Properties, expectedMessage: String): Unit = {
+    val exception = assertThrows(classOf[KafkaException], () => new LineMessageReader().init(null, props))
+    assertEquals(
+      expectedMessage,
+      exception.getMessage
+    )
+  }
+
+  @Test
+  def testIgnoreErrorInInput(): Unit = {
+    val input =
+      "headerKey0.0:headerValue0.0,headerKey0.1:headerValue0.1[MISSING-DELIMITER]key0\tvalue0\n" +
+      "headerKey1.0:headerValue1.0\tkey1\tvalue1"
+
+    val props = defaultTestProps
+    props.put("ignore.error", "true")
+
+    val invalidRecord: ProducerRecord[String, String] =
+      record(
+        null,
+        "value0",
+        List("headerKey0.0" -> "headerValue0.0", "headerKey0.1" -> "headerValue0.1[MISSING-DELIMITER]key0")
+      )
+
+    val validRecord = record("key1", "value1", List("headerKey1.0" -> "headerValue1.0"))
+
+    runTest(props, input, invalidRecord, validRecord)
+  }
+
+  @Test
+  def testMalformedHeader(): Unit = {
+    val lineReader = new LineMessageReader()
+    val input = "key-val\tkey0\tvalue0\n"
+    lineReader.init(new ByteArrayInputStream(input.getBytes), defaultTestProps)
+
+    val expectedException = assertThrows(classOf[KafkaException], () => lineReader.readMessage())
+
+    assertEquals(
+      "No header key separator found in pair 'key-val' on line number 1",
+      expectedException.getMessage
+    )
+  }
+
+  @Test
+  def testMalformedHeaderIgnoreError(): Unit = {
+    val input = "key-val\tkey0\tvalue0\n"
+
+    val props = defaultTestProps
+    props.put("ignore.error", "true")
+
+    val expected = record("key0", "value0", List("key-val" -> null))
+
+    runTest(props, input, expected)
+  }
+
+  def runTest(props: Properties, input: String, expectedRecords: ProducerRecord[String, String]*): Unit = {
+    val lineReader = new LineMessageReader()
+    lineReader.init(new ByteArrayInputStream(input.getBytes), props)
+    expectedRecords.foreach(r => assertEquality(r, lineReader.readMessage()))
+  }
+
+  //  The equality method of ProducerRecord compares memory references for the header iterator, this is why this custom equality check is used.
+  private def assertEquality[K, V](expected: ProducerRecord[K, V], actual: ProducerRecord[Array[Byte], Array[Byte]]): Unit = {
+    assertEquals(expected.key(), if (actual.key() == null) null else new String(actual.key()))
+    assertEquals(expected.value(), if (actual.value() == null) null else new String(actual.value()))
+    assertEquals(expected.headers().toArray.toList, actual.headers().toArray.toList)
```

Comment:
nit: We could omit the parenthesis for all the getters here (e.g. `key`, `value`, `headers`). We usually don't put them for getters in our code base. There might be a few other cases in other method. I haven't checked them all.